IndexOutOFBoundsException when resolving with multiple varargs methods available.

I found this issue when resolving code containing calls to log4j but have now recreated the issue with some much simpler code that has no dependencies on other libraries:

```
import java.util.function.Supplier;

public class AParserTest {

	public void testMethod() {
		doSomething("abc");
	}
	
	private void doSomething(String string, Supplier<?>... supplier) {
		
	}

	private void doSomething(String string, Object... object) {
		
	}
}
```
The parsing code I used was very simple, here is the main class:
```
public class Parser {
	
	private static final String PATH_TO_SOURCE = "...";

	private static final Logger LOGGER = LogManager.getLogger();
	public static void main(String[] args) {
		
		try {
			TypeSolver typeSolver = new CombinedTypeSolver(
					new ReflectionTypeSolver());
			JavaSymbolSolver symbolSolver = new JavaSymbolSolver(typeSolver);
			
			StaticJavaParser.getConfiguration().setSymbolResolver(symbolSolver);

			CompilationUnit cu = StaticJavaParser.parse(New File(PATH_TO_SOURCE));
				
			List<String> messages = new ArrayList<>();
			VoidVisitor<List<String>> collector = new Collector();
			collector.visit(cu, messages);
		} catch (Exception ex) {
			String message = "Exception thrown";
			LOGGER.error(message, ex);
			throw new RuntimeException(message, ex);
		}
	}
}
```
And the Collector:
```
public class Collector extends VoidVisitorAdapter<List<String>> {

	private static final Logger LOGGER = LogManager.getLogger();
	
	@Override
	public void visit(MethodCallExpr mce, List<String> collector) {
		
		super.visit(mce, collector);

		try {
			String methodQualifiedMethodName = mce.resolve().getQualifiedName();
			LOGGER.info("Method: {}", methodQualifiedMethodName);
			
		} catch (UnsolvedSymbolException ex) {
			LOGGER.warn("Skipping method {}, not found in core or referenced JAR(s)", mce.getName());
		}
		
	}
}
```
The execution of the above yields an IndexOutOfBoundsException:
```
[ERROR] 2021-01-08 08:43:55.900 [main] Parser - Exception thrown
java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
	at java.util.LinkedList.checkElementIndex(LinkedList.java:555) ~[?:1.8.0_151]
	at java.util.LinkedList.get(LinkedList.java:476) ~[?:1.8.0_151]
	at com.github.javaparser.symbolsolver.resolution.MethodResolutionLogic.isExactMatch(MethodResolutionLogic.java:597) ~[javaparser-symbol-solver-core-3.18.0.jar:?]
	at com.github.javaparser.symbolsolver.resolution.MethodResolutionLogic.findMostApplicable(MethodResolutionLogic.java:583) ~[javaparser-symbol-solver-core-3.18.0.jar:?]
	at com.github.javaparser.symbolsolver.resolution.MethodResolutionLogic.findMostApplicable(MethodResolutionLogic.java:501) ~[javaparser-symbol-solver-core-3.18.0.jar:?]
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.JavaParserTypeDeclarationAdapter.solveMethod(JavaParserTypeDeclarationAdapter.java:206) ~[javaparser-symbol-solver-core-3.18.0.jar:?]
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.ClassOrInterfaceDeclarationContext.solveMethod(ClassOrInterfaceDeclarationContext.java:126) ~[javaparser-symbol-solver-core-3.18.0.jar:?]
	at com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParserClassDeclaration.solveMethod(JavaParserClassDeclaration.java:326) ~[javaparser-symbol-solver-core-3.18.0.jar:?]
```

The javaparser is expecting to find multiple arguments in the call to the doSomething method - But throws the exception when it tries to get the second argument, because there isn't one. 

The issue only occurs when both doSomething methods are present in the AParserTest class (the Object and Supplier) - If one or the other is removed, the code parses without issue. 

I'm using Java 1.8.0_151 (yes, I know it's old), and this Maven dependency for everything javaparser related:
```
<dependency>
   <groupId>com.github.javaparser</groupId>
   <artifactId>javaparser-symbol-solver-core</artifactId>
   <version>3.18.0</version>
</dependency>
```

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/MethodResolutionLogic.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/MethodResolutionLogic.java
@@ -623,6 +623,7 @@ public class MethodResolutionLogic {
         final int numberOfArgs = argumentTypes.size();
         final ResolvedType lastArgType = numberOfArgs > 0 ? argumentTypes.get(numberOfArgs - 1) : null;
         final boolean isLastArgArray = lastArgType != null && lastArgType.isArray();
+        int omittedArgs = 0;
 
         // If one method declaration has exactly the correct amount of parameters and is not variadic then it is always
         // preferred to a declaration that is variadic (and hence possibly also has a different amount of parameters).
@@ -633,12 +634,22 @@ public class MethodResolutionLogic {
                 !isLastArgArray))) {
             return false;
         }
+        
+        // If both methods are variadic but the calling method omits any varArgs, bump the omitted args to
+        // ensure the varargs type is considered when determining which method is more specific
+        if (aVariadic && bVariadic && aNumberOfParams == bNumberOfParams && numberOfArgs == aNumberOfParams - 1) {
+        	omittedArgs++;
+        }
 
         // Either both methods are variadic or neither is. So we must compare the parameter types.
-        for (int i = 0; i < numberOfArgs; i++) {
+        for (int i = 0; i < numberOfArgs + omittedArgs; i++) {
             ResolvedType paramTypeA = getMethodsExplicitAndVariadicParameterType(methodA, i);
             ResolvedType paramTypeB = getMethodsExplicitAndVariadicParameterType(methodB, i);
-            ResolvedType argType = argumentTypes.get(i);
+            
+            ResolvedType argType = null;
+            if (i < argumentTypes.size()) {
+            	argType = argumentTypes.get(i);
+            }
 
             // Safety: if a type is null it means a signature with too few parameters managed to get to this point.
             // This should not happen but it also means that this signature is immediately disqualified.
@@ -652,12 +663,14 @@ public class MethodResolutionLogic {
             // The method call will call foo(long), as it requires a widening primitive conversion from int to long
             // instead of a boxing conversion from int to Integer. See JLS §15.12.2.
             // This is what we check here.
-            else if (paramTypeA.isPrimitive() == argType.isPrimitive() &&
+            else if (argType != null &&
+            		paramTypeA.isPrimitive() == argType.isPrimitive() &&
                     paramTypeB.isPrimitive() != argType.isPrimitive() &&
                     paramTypeA.isAssignableBy(argType)) {
 
                 return true;
-            } else if (paramTypeB.isPrimitive() == argType.isPrimitive() &&
+            } else if (argType != null &&
+            		paramTypeB.isPrimitive() == argType.isPrimitive() &&
                     paramTypeA.isPrimitive() != argType.isPrimitive() &&
                     paramTypeB.isAssignableBy(argType)) {
 

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue3028Test.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue3028Test.java
@@ -1,0 +1,36 @@
+package com.github.javaparser.symbolsolver;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+import com.github.javaparser.StaticJavaParser;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.expr.MethodCallExpr;
+import com.github.javaparser.resolution.declarations.ResolvedMethodDeclaration;
+import com.github.javaparser.symbolsolver.javaparser.Navigator;
+import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
+import com.github.javaparser.symbolsolver.resolution.AbstractResolutionTest;
+import com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
+import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;
+
+public class Issue3028Test extends AbstractResolutionTest {
+
+    @Test
+    void varArgsIssue() {
+    	
+    	TypeSolver typeSolver = new CombinedTypeSolver(
+				new ReflectionTypeSolver());
+		JavaSymbolSolver symbolSolver = new JavaSymbolSolver(typeSolver);
+		
+		StaticJavaParser.getConfiguration().setSymbolResolver(symbolSolver);
+		
+        CompilationUnit cu = parseSample("Issue3028");
+        
+        final MethodCallExpr mce = Navigator.findMethodCall(cu, "doSomething").get();
+        ResolvedMethodDeclaration rmd = mce.resolve();
+        assertEquals("AParserTest.doSomething(java.lang.String, java.util.function.Supplier<?>...)", rmd.getQualifiedSignature());
+        
+    }
+    
+}

--- a/javaparser-symbol-solver-testing/src/test/resources/Issue3028.java.txt
+++ b/javaparser-symbol-solver-testing/src/test/resources/Issue3028.java.txt
@@ -1,0 +1,16 @@
+import java.util.function.Supplier;
+
+public class AParserTest {
+
+	public void testMethod() {
+		doSomething("abc");
+	}
+	
+	private void doSomething(String string, Supplier<?>... supplier) {
+		
+	}
+
+	private void doSomething(String string, Object... object) {
+		
+	}
+}

